### ISSUE TITLE
When I run shared memory within a window service, Unable to read from applications as console application 

### ISSUE THREAD

**Issue by vinaydwivedi** [spazzarama/SharedMemory#53]
private SharedMemory.SharedArray<MarketTickS> _buffer = _buffer = new SharedArray<MarketTickS>(@"Global\MARKETTICK", 200000); from within window service.

Try creating 
SharedMemory.SharedArray<MarketTickS> _reader = new SharedArray<MarketTickS>(@"Global\MARKETTICK");

I Get "Access to the path 'Global\\MARKETTICK_evt_write' is denied"

When I create the file without Global Prefix i.e.
private SharedMemory.SharedArray<MarketTickS> _buffer = _buffer = new SharedArray<MarketTickS>("MARKETTICK", 200000); from within window service.

from Console
SharedMemory.SharedArray<MarketTickS> _reader = new SharedArray<MarketTickS>(@"MARKETTICK");

I get file not found error.

**Comment by spazzarama** [comment]
What is the window service authenticating as? I would say you need to change this.

**Comment by vinaydwivedi** [comment]
Window service is authenticated as Network service, tried running it under local service account as well, doesn't work.

**Comment by spazzarama** [comment]
I believe the service must be running using an account that has the Local Security Policy “Create global objects” allowed.

**Comment by vinaydwivedi** [comment]
Yes, please see below, Just to be sure, if this is not permission issue, I granted create "create global objects" to everyone.

![image](https://user-images.githubusercontent.com/9931326/91972124-daea6500-ed37-11ea-8b51-b5f75b9b489e.png)

**Comment by vinaydwivedi** [comment]
Could following be required, don't see this being set. 

var security = new MemoryMappedFileSecurity();
security.AddAccessRule(new AccessRule<MemoryMappedFileRights>("everyone", MemoryMappedFileRights.FullControl,AccessControlType.Allow));

**Comment by spazzarama** [comment]
Probably for the client to connect but that does not explain why you can’t create the initial MMF.

**Comment by vinaydwivedi** [comment]
Both the issue were from client side. I have switched to Managed version of MemoryMapped files, and able to work as expected with security MemoryMappedFileSecurity added.

**Comment by spazzarama** [comment]
Ah I thought the issue was creating not opening existing. Anyway, sounds like that is the issue, thanks.

**Comment by lukiluki123** [comment]
Hey, vinaydwivedi. 
Can you explain your solution more detailed?
I have similar issue but with CircularBuffer, between service (net7) and consol (net35)

**Comment by wsloth** [comment]
Hey all, bumping this issue thread again because we are also running into this issue.

SharedMemory via RpcBuffer works fine when the host & client are console applications, but no longer when the host is a Windows Service. 

I tried the `Global\Name` approach, but this is not a feasible solution for us as we are deploying to hundreds of machines and would rather not modify permissions for all users on these machines just for sharing a mutex lock.

I also tried cloning the project, adding `MutexSecurity` as a parameter to `RpcBuffer` and using it to initialize the Mutex. However, I still ran into the same issues (probably due to a lack of knowledge about how to properly configure this).

Is there anyone in this thread who is willing to post their solution?

**Comment by spazzarama** [comment]
I haven’t had a chance to look at this in a lot of detail yet, but the ‘MutexSecurity’ looks like the correct approach to me.
